File: java/src/main/java/com/bparent/workshop/unittests/bean/RegularFrame.java
package com.bparent.workshop.unittests.bean;

import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@NoArgsConstructor
public class RegularFrame extends Frame {

    public void registerShot(Integer shotValue) {
        if (!canAcceptShot()) {
            throw new IllegalStateException("This frame cannot accept shots anymore");
        }
        if (shotValue > 10) {
            throw new IllegalArgumentException("You cannot have a better score than 10 for one shot");
        }
        if (this.firstShot != null && this.firstShot + shotValue > 10) {
            throw new IllegalArgumentException("The sum of 2 shots cannot be > 10");
        }

        if (this.firstShot == null) {
            this.firstShot = shotValue;
        } else {
            this.secondShot = shotValue;
        }
    }

    public List<Integer> getAllShots() {
        return Stream.of(
                this.firstShot,
                this.secondShot
        ).filter(Objects::nonNull).collect(Collectors.toList());
    }

    public String toString() {
        return String.format("[%s | %s]", formatShot(this.firstShot), formatShot(this.secondShot));
    }

    public boolean canAcceptShot() {
        return this.firstShot == null ||
                (this.firstShot != 10 && this.secondShot == null);
    }

    public boolean isSpare() {
        return this.firstShot != null && this.secondShot != null &&
                this.firstShot + this.secondShot == 10;
    }

    public boolean isStrike() {
        return this.firstShot != null && this.firstShot == 10;
    }

}

File: js/js/regularFrame.js
export default class RegularFrame {
  constructor(firstShot = null, secondShot = null, score = null) {
    this.firstShot = firstShot;
    this.secondShot = secondShot;
    this.score = score;
  }

  canAcceptShot() {
    return !this.firstShot || (this.firstShot !== 10 && !this.secondShot);
  }

  registerShot(shotValue) {
    const value = +shotValue;
    if (!this.canAcceptShot()) {
      throw new Error("This frame cannot accept shots anymore");
    }
    if (value > 10) {
      throw new Error("You cannot have a better score than 10 for one shot");
    }
    if (this.firstShot !== null && this.firstShot + value > 10) {
      throw new Error("The sum of 2 shots cannot be > 10. First shot = " + this.firstShot + ". New shot value : " + value);
    }

    if (this.firstShot === null) {
      this.firstShot = value;
    } else {
      this.secondShot = value;
    }
  }

  isSpare() {
    return this.firstShot !== null && this.secondShot !== null &&
      this.firstShot + this.secondShot === 10;
  }

  isStrike() {
    return this.firstShot !== null && this.firstShot === 10;
  }

  getAllShots() {
    return [ this.firstShot, this.secondShot ].filter(shot => shot !== null);
  }


  render(index) {
    return `
    <li>
        <span class="frame-score">${ this.score || "-" }</span>
        
        <div class="frame-shots">
            <div class="frame-shot-index">${ index }</div>
            <div class="frame-shot-first">${ this.firstShot || "-" }</div>
            <div class="frame-shot-second">${ this.secondShot || "-" }</div>
        </div>
    </li>`;
  }
}
